### ISSUE TITLE
Unable to browse by artists in Subsonic clients since recent commit

### ISSUE THREAD

**Issue by JoeDat** [ampache/ampache#2853]
## Description
Since [c6d0a0e](https://github.com/ampache/ampache/commit/c6d0a0e9ef09f1bd626ad703d57ea7d40bb4daa0) I am no longer able to browse by artist via Subsonic clients (currently with DSub and Sublime Music)

### Describe the bug
When attempting to pull a list of artists via Subsonic clients, no data is returned.  If I rollback to the [parent](https://github.com/ampache/ampache/commit/517a2993da6860c9ff4be1c8c7a624853a06ef4c), problem is gone.  Pulling data via other views (Album, Playlist, etc.) appear unaffected.  I've tried adding/cleaning/updating my catalog to no avail.

### To reproduce

DSub:
1. Click 'Library'
2. DSub will retry 4 times before returning 'No Data'

Sublime
1. Click the Artists button at the top for Artists view
2. Click the refresh button in the left artist column
3. Column will spin a loading icon, before returning 'Error attempting to load artists'

### Expected behavior
Subsonic clients should return artist data

## Environment

* Ampache version: 5.0.0
* Web server + version: Apache 2.4.46
* Server operating system: Fedora 33
* Client operating system: Android and Fedora 33 Workstation

### Client type

* SubSonic

## Config

Since I'm making no changes to my configuration between commits, figuring this isn't needed.  However, can certainly supply if I'm wrong!!

## Logs

Only log entry when in debug mode shows a successful execution of the getArtists API call
`
2021-03-26T10:24:47-04:00 [admin] (Ampache\Module\Api\Subsonic_Xml_Data) -> API success in function getartists-1.13.0`

**Comment by lachlan-00** [comment]
i'll check it out, i use both. it's supposed to get those counts when they're missig but maybe it hasn't filled.

do a catalog add as well in the mean time and the server will fill the new columns

**Comment by JoeDat** [comment]
I actually tried the catalog add a few times.  I saw it mentioned in the DB update view and suspected this was related.  Tried just an add, then a clean/add/update.  No errors are written to the logs.  I can try to completely remove and re-add my catalog as well later today.  Maybe that'll do it.

**Comment by JoeDat** [comment]
So, good news is that worked.  I didn't exactly remove and re-add the catalog, but moved all of my files out of the catalog, cleaned, moved them back, then added.  I can browse by artist now in both Subsonic clients.

However, I'm now seeing errors relating to the Count being thrown in the logs when browsing in Sublime Music only.  No such errors in DSub (doesn't appear to request it).  Despite this, Sublime is showing accurate song/album counts.  I continue to see this after updating to the latest commit.  

Of course, this is a nuisance only.  If it's something worth looking into it, I'll help anyway I can.  But, if not, I'm good with closing this one out.  

`2021-03-29T09:35:08-04:00 [admin] (Ampache\Module\Api\Subsonic_Xml_Data) -> API success in function getartistinfo-1.13.0
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] count(): Parameter must be an array or an object that implements Countable in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(351)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] Invalid argument supplied for foreach() in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(401)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] count(): Parameter must be an array or an object that implements Countable in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(351)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] Invalid argument supplied for foreach() in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(401)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] count(): Parameter must be an array or an object that implements Countable in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(351)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] Invalid argument supplied for foreach() in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(401)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] count(): Parameter must be an array or an object that implements Countable in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(351)
2021-03-29T09:35:08-04:00 [ampache] (log.lib) -> [Runtime Error] Invalid argument supplied for foreach() in file /mnt/data/srv/www/ampache/src/Module/Api/Subsonic_Api.php(401)
2021-03-29T09:35:09-04:00 [] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Image\ShowAction" for action "show"
`

**Comment by lachlan-00** [comment]
sublime is json format and that's in the conversion function.
![image](https://user-images.githubusercontent.com/1305249/112907256-e4527900-9130-11eb-8081-1b0252a2ae30.png)

Just looking i get the same errors but can browse okay. it seems like this falls through to the right spot but needs to be fixed.

**Comment by lachlan-00** [comment]
you've also pointed me to why artist art is failing in sublime! i guess subsonic stores the art in a folder somewhere and returns that?

8bed4d655 fixes artist art in sublime as well as skipping strings in conversion checks for json makes sublime a lot better how that i know why the art was failing.

**Comment by JoeDat** [comment]
Interesting that I haven't been having artist art issues in Sublime ¯\_(ツ)_/¯

Everything is good to go.  Think I can close this out.  Thanks!!

**Comment by JoeDat** [comment]
So, this problem has actually returned for me.  The previous steps I took to completely remove my music, clean, then re-add, are not working.  And, as with the initial report, resetting my local repo back to [517a299](https://github.com/ampache/ampache/commit/517a2993da6860c9ff4be1c8c7a624853a06ef4c) is resolving.  This time, I dug a little deeper and can see gateway timeouts being thrown in the logs.  Plugging the URL directly into the browser produces the same timeout.

```
<IP> - - [14/Apr/2021:10:34:52 -0400] "GET /rest/getArtists.view?u=admin&p=enc:<encoded_pass>&v=1.2.0&c=DSub HTTP/1.1" 504 247
<IP> - - [14/Apr/2021:10:35:09 -0400] "GET /rest/getArtists.view?u=admin&p=enc:<encoded_pass>&v=1.2.0&c=DSub HTTP/1.1" 504 247
```
All other API calls are successful.  I can browser via album, calling a stream works.  Only an issue with getArtists.  Not sure how I can debug this further.

**Comment by JoeDat** [comment]
Monitored my database when refreshing my artist view.  It looks like it triggers some kind of process that does a mass update on the artist table, rewriting all the last_update fields with current time.  I'm wondering if this is causing some kind of delay that forces the call from the Subsonic client to timeout.  These updates do not happen on the [517a299](https://github.com/ampache/ampache/commit/517a2993da6860c9ff4be1c8c7a624853a06ef4cl)

![image](https://user-images.githubusercontent.com/8648894/114791888-41674500-9d55-11eb-8ecf-c7b5774c9414.png)

**Comment by lachlan-00** [comment]
in the next commit after that the database adds a column for the counts.
https://github.com/ampache/ampache/commit/c6d0a0e9ef09f1bd626ad703d57ea7d40bb4daa0

Whats probably happening is that your DB hasn't filled all the columns yet.

update back and run a catalog add on the DB to fill the columns. this should fill the data and speed things up.

This column was added to stop having to calculate the counts each call to artists so it should be faster for you once the add is completed.

**Comment by JoeDat** [comment]
Didn't help.  Running the add did perform another mass update on all the last_update counts of the artists in the database, but both DSub and Sublime are still re-triggering this process when refreshing artist view.  I suppose that's expected, and maybe this change has just exposed a severe inefficiency in the performance of my DB instance.  I have ~890 distinct artists that it has to go thru with each artist scan and I don't have artists with _NULL_ counts in any of the count columns.  I'll see if there's anything I can do to speed things up.

**Comment by lachlan-00** [comment]
after the initial counts it shouldn't do it each time it loads that's def going to slow you down.

**Comment by lachlan-00** [comment]
yeah i see it doing the same for me, it doesn't need to so we'll stop that

**Comment by JoeDat** [comment]
Just following up on this issue.  Totally understand waiting on it, I'm running a dev version of the app after all.  It's just a complete showstopper for me, I'm assuming because of the number of artists I'm iterating through on every lookup.

**Comment by lachlan-00** [comment]
i've done some testing recently and i can see it happening in develop only so it's at least not in master yet. will be working on it next after the next release as it is annoying!

**Comment by lachlan-00** [comment]
well there's the issue....
```
2021-04-30T14:29:02+10:00 [user] (ajax.server) -> Called for page: {search}
2021-04-30T14:29:03+10:00 [user] (Ampache\Repository\Model\Artist) -> album_count `1` != current count `1`
2021-04-30T14:29:03+10:00 [user] (Ampache\Repository\Model\Artist) -> group_count `1` != current count `1`
2021-04-30T14:29:03+10:00 [user] (Ampache\Repository\Model\Artist) -> song_count `11` != current count `11`
2021-04-30T14:29:03+10:00 [user] (Ampache\Repository\Model\Artist) -> album_count `1` != current count `1`
2021-04-30T14:29:03+10:00 [user] (Ampache\Repository\Model\Artist) -> group_count `1` != current count `1`
2021-04-30T14:29:03+10:00 [user] (Ampache\Repository\Model\Artist) -> song_count `11` != current count `11`

```

**Comment by lachlan-00** [comment]
@JoeDat I can get 14k artists in 3 seconds now. just need to clean up the code first

**Comment by lachlan-00** [comment]
try out 67e5152cf

**Comment by JoeDat** [comment]
FIXED!!  It's as snappy as it used to be.  Awesome work!

**Comment by JoeDat** [comment]
Sorry, I keep jumping the gun an closing this out too early.  I'm normally a little more thorough than that!

So, what made me think this was good was because DSub was loading snappy again, which is awesome.  However, I tried Sublime Music, and there's no improvement.  I looked at DSub again and noticed that while I was troubleshooting this previously, I unchecked 'Browse by Tags', which resorts to folder browsing.  This was also timing out until now (hence, making me think everything was fixed!)  If I check this option, I'm back to seeing the timeouts.  It seems that browsing by folders makes a successful API call to getindexes.  Sublime Music also calls getindexes with the Browse option, but I get no data back.   I honestly don't recall trying Browse in Sublime Music before now.  

I can't argue with having partial functionality back in DSub.  I can live with folder browsing, and I use DSub way more than Sublime!

To recap:

DSub:
- Uncheck Browse by Tags in the Server config:  Loads instantly.
- Check Browse by Tags:  Times out

Sublime Music:
- Click 'Artists': Times out
- Click 'Browse': No time out, but no data returned.

All API calls are successful in the logs.

**Comment by lachlan-00** [comment]
Depending on what they want to load I can speed those ones up too.

Will check again later

**Comment by lachlan-00** [comment]
just looking, the getartists calls doesn't really add much more and it's all in the DB so that should be easy

**Comment by lachlan-00** [comment]
@JoeDat 0e35d5265 fixes getartists.

there seems to be an error in sublime when clicking browse, i can't see any calls made. need to dig a bit more

**Comment by JoeDat** [comment]
Weird.  I see a call to getindexes.  Doesn't throw an error in Sublime.  I just get nothing back.  It's unfrozen.  

`2021-05-03T20:10:49-04:00 [admin] (Ampache\Module\Api\Subsonic_Xml_Data) -> API success in function getindexes-1.13.0`

![image](https://user-images.githubusercontent.com/8648894/116948126-8f80b180-ac4c-11eb-93bc-46e432db7068.png)

**Comment by lachlan-00** [comment]
sublime is pretty strict on the api compliance so i'm sure there are some more hidden errors. i need to make a smaller library as it's loading okay now but seems to cache / read through a lot of data to do it.

**Comment by JoeDat** [comment]
Just revisiting this bug.  I know there have been a ton of commits in recent months, and right now, Sublime is working normally.  Browse and Artists are responding perfectly.  And for the record, so is DSub.  

Thinking we can close this one out permanently!
